libs/mock-block-dock/src/datastore/traverse.ts

```diff
@@ -1,284 +1,543 @@
-              datastore,
-              entityId,
-            )) {
-              const {
-                metadata: {
-                  recordId: {
-                    entityId: outgoingLinkEntityId,
-                    editionId: edgeTimestamp,
-                  },
-                },
-              } = outgoingLinkEntity;
+  for (const [edgeKind, depthsPerDirection] of typedEntries(
+    currentTraversalDepths,
+  )) {
+    for (const [direction, depth] of typedEntries(depthsPerDirection)) {
+      if (depth < 1) {
+        continue;
+      }
 
-              const outgoingLinkEdge: OutgoingLinkEdge = {
-                kind: "HAS_LEFT_ENTITY",
-                reversed: true,
-                rightEndpoint: {
-                  baseId: outgoingLinkEntityId,
-                  timestamp: edgeTimestamp,
-                },
-              };
+      for (const neighborVertex of getNeighbors(
+        traversalSubgraph,
+        datastore,
+        element,
+        /** @todo - this will be unergonomic as soon as there are more supported edge kinds */
+        edgeKind === "hasLeftEntity" ? "HAS_LEFT_ENTITY" : "HAS_RIGHT_ENTITY",
+        direction === "incoming",
+        interval,
+      )) {
+        let newIntersection: NonNullTimeInterval | null;
+        let neighborVertexId: GraphElementVertexId;
 
-              addKnowledgeGraphEdgeToSubgraphByMutation(
-                traversalSubgraph,
-                entityId,
-                edgeTimestamp,
-                outgoingLinkEdge,
-              );
-
-              /** @todo - This is temporary, and wrong */
-              toResolve.insert(
-                {
-                  baseId: outgoingLinkEntity.metadata.recordId.entityId,
-                  revisionId: outgoingLinkEntity.metadata.recordId.editionId,
-                },
-                {
-                  [edgeKind]: { incoming: depths.incoming - 1 },
-                },
-              );
-            }
-          }
-          if (depths.outgoing) {
-            if (
-              "linkData" in element.inner &&
-              element.inner.linkData?.leftEntityId !== undefined &&
-              element.inner.linkData?.rightEntityId !== undefined
-            ) {
-              // get left entity for link entity and insert edges
-              const leftEntity = getLeftEntityForLinkEntity(
-                datastore,
-                entityId,
-              );
-              const {
-                metadata: {
-                  recordId: {
-                    entityId: leftEntityId,
-                    editionId: edgeTimestamp,
-                  },
-                },
-              } = leftEntity;
+        if (isEntityVertex(neighborVertex)) {
+          // get from temporal data of the neighbor vertex
+          const entityValidInterval =
+            neighborVertex.inner.metadata.temporalVersioning[
+              traversalSubgraph.temporalAxes.resolved.pinned.axis
+            ];
```

Comment:
I cannot really follow this code. What is `entityValidInterval` in this context? Why you're indexing via the `pinned` axis?